File: experimental/verify_mca_full_lift_fixed_cutoff_boundary_stack_v1.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

typedef __int128 i128;

enum {
    R = 1048576,
    D = 67448,
    K = 6,
    N = R + K,
    M = D + K,
    C = K - 1,
    BUDGET = 16777215,
    CUTOFF = 65200,
};

static int64_t caps[CUTOFF + 1];

static void reject(const char *what, int e, int h) {
    fprintf(stderr, "REJECT %s e=%d h=%d\n", what, e, h);
    exit(1);
}

static int64_t raw_cap(int e, int h) {
    int64_t n = N - e;
    int64_t agreement = M - h;
    int64_t johnson = agreement * agreement - n * C;
    if (johnson > 0) {
        return (int64_t)(((i128)n * (agreement - C)) / johnson);
    }
    int64_t gap = -johnson;
    int64_t balance = 2 * agreement * agreement - n * C;
    int64_t tangent = (n - agreement) * (n - agreement) - (n - 1) * gap;
    if (balance < 0 || tangent <= 0) reject("prefix cap", e, h);
    i128 numerator = (i128)(n - 1) * n * n * (agreement - C);
    i128 denominator = (i128)agreement * tangent;
    return (int64_t)(numerator / denominator);
}

static int64_t prefix(int e) {
    caps[0] = 0;
    for (int h = 1; h <= CUTOFF; ++h) caps[h] = raw_cap(e, h);
    for (int h = CUTOFF - 1; h >= 1; --h) {
        if (caps[h + 1] < caps[h]) caps[h] = caps[h + 1];
    }
    int64_t total = 0;
    for (int h = 1; h <= CUTOFF; ++h) {
        total += (caps[h] - caps[h - 1]) * (e / h);
    }
    return total;
}

static int64_t boundary_cap(int e, int h) {
    int64_t n = N - e;
    int64_t inside = 2LL * h - e;
    int64_t denominator = inside * inside - (int64_t)e * C;
    int64_t outside = M - h;
    if (2 * h <= e || inside <= 0 || denominator <= 0 ||
            !(n > outside && outside > C)) {
        reject("boundary cap", e, h);
    }
    int64_t classes = (int64_t)(((i128)e * (inside - C)) / denominator);
    int64_t line = (n - C) / (outside - C);
    return 1 + classes * (line - 1);
}

static void checkpoint(int e, int64_t p, int64_t stack, int64_t forcing,
                       int64_t threshold, int64_t core, int64_t low,
                       int64_t bound) {
    if (e == 98232 && !(p == 1381829 && stack == 391210 &&
            forcing == 1773039 && threshold == 15004177)) {
        reject("first checkpoint", e, 0);
    }
    if (e == 101149 && !(p == 1422377 && stack == 14327810 &&
            forcing == 15750187 && threshold == 1027029)) {
        reject("last direct checkpoint", e, 0);
    }
    if (e == 101150 && !(p == 1422391 && stack == 14530797 &&
            forcing == 15953188 && threshold == 824028 &&
            core == 67453 && low == 28 && bound == 3813329)) {
        reject("first absorption checkpoint", e, 0);
    }
    if (e == 101155 && !(p == 1422461 && stack == 15244572 &&
            forcing == 16667033 && threshold == 110183 &&
            core == 67446 && low == 28 && bound == 3813469)) {
        reject("endpoint checkpoint", e, 0);
    }
    if (e == 101156 && !(p == 1422475 && stack == 15528748 &&
            forcing == 16951223 && threshold == -174007)) {
        reject("adjacent checkpoint", e, 0);
    }
}

int main(void) {
    int direct = 0, absorption = 0, wall = 0;
    for (int e = 98232; e <= 101156; ++e) {
        int s = (e - K) / 3;
        int H = e - s - 1;
        int64_t p = prefix(e);
        int64_t stack = 0;
        for (int h = CUTOFF + 1; h <= H; ++h) {
            stack += boundary_cap(e, h);
        }
        int64_t forcing = p + stack;
        int64_t threshold = BUDGET - forcing + 1;
        int64_t core = -1, low = -1, bound = -1;

        if (forcing + (N - M + 1) <= BUDGET) {
            if (e > 101149) reject("late direct branch", e, H);
            ++direct;
        } else if (threshold >= 2) {
            core = (threshold * (int64_t)M - N + threshold - 2)
                   / (threshold - 1);
            int64_t inside = core - C;
            int64_t sync = e - inside + K;
            int64_t agreement = M - sync + 1;
            int64_t n = N - e;
            int64_t denominator = agreement * agreement - n * C;
            if (denominator <= 0) reject("low Johnson", e, H);
            low = (n * (agreement - C)) / denominator;
            bound = e * low + (N - M + 1);
            if (e > 101155 || bound >= BUDGET) {
                reject("absorption payment", e, H);
            }
            ++absorption;
        } else {
            if (e != 101156) reject("early wall", e, H);
            ++wall;
        }
        checkpoint(e, p, stack, forcing, threshold, core, low, bound);
    }
    if (direct != 2918 || absorption != 6 || wall != 1) {
        reject("branch census", 0, 0);
    }
    puts("MCA_FULL_LIFT_FIXED_CUTOFF_BOUNDARY_STACK_V1_SCAN_PASS "
         "paid=2924 direct=2918 absorption=6 adjacent_wall=1");
    return 0;
}

File: experimental/verify_mca_full_lift_fixed_cutoff_boundary_stack_v1.py
#!/usr/bin/env python3
"""Verify endpoints of the M31 fixed-cutoff boundary-stack payment."""

from __future__ import annotations

import copy


R, D, K, BUDGET, CUTOFF = 1048576, 67448, 6, 16777215, 65200
N, M, C = R + K, D + K, K - 1
EXPECTED = {
    98232: (1381829, 391210, 1773039, 15004177, "direct"),
    101149: (1422377, 14327810, 15750187, 1027029, "direct"),
    101150: (1422391, 14530797, 15953188, 824028,
             "absorption", 67453, 28, 3813329),
    101155: (1422461, 15244572, 16667033, 110183,
             "absorption", 67446, 28, 3813469),
    101156: (1422475, 15528748, 16951223, -174007,
             "fixed_cutoff_wall"),
}


class Reject(ValueError):
    pass


def cap(e: int, h: int) -> int:
    n, agreement = N - e, M - h
    johnson = agreement * agreement - n * C
    if johnson > 0:
        return n * (agreement - C) // johnson
    gap = -johnson
    balance = 2 * agreement * agreement - n * C
    tangent = (n - agreement) ** 2 - (n - 1) * gap
    if balance < 0 or tangent <= 0:
        raise Reject("undefined prefix cap")
    return ((n - 1) * n * n * (agreement - C)
            // (agreement * tangent))


def prefix(e: int) -> int:
    values = [0] + [cap(e, h) for h in range(1, CUTOFF + 1)]
    for h in range(CUTOFF - 1, 0, -1):
        values[h] = min(values[h], values[h + 1])
    return sum((values[h] - values[h - 1]) * (e // h)
               for h in range(1, CUTOFF + 1))


def record(e: int):
    s, _ = divmod(e - K, 3)
    H = e - s - 1
    p = prefix(e)
    stack = 0
    for h in range(CUTOFF + 1, H + 1):
        A, n, outside = 2 * h - e, N - e, M - h
        denominator = A * A - e * C
        if (2 * h <= e or denominator <= 0 or not (n > outside > C)):
            raise Reject("boundary guard")
        classes = e * (A - C) // denominator
        line = (n - C) // (outside - C)
        stack += 1 + classes * (line - 1)
    forcing = p + stack
    threshold = BUDGET - forcing + 1
    if forcing + (N - M + 1) <= BUDGET:
        return (p, stack, forcing, threshold, "direct")
    if threshold < 2:
        return (p, stack, forcing, threshold, "fixed_cutoff_wall")
    core = (threshold * M - N + threshold - 2) // (threshold - 1)
    inside = core - C
    sync = e - inside + K
    agreement = M - sync + 1
    n = N - e
    denominator = agreement * agreement - n * C
    if denominator <= 0:
        raise Reject("low Johnson")
    low = n * (agreement - C) // denominator
    bound = e * low + (N - M + 1)
    return (p, stack, forcing, threshold, "absorption", core, low, bound)


def validate(expected):
    for e, row in expected.items():
        if record(e) != row:
            raise Reject(f"endpoint {e}")
    if 101155 - 98232 + 1 != 2924:
        raise Reject("census")
    if EXPECTED[101156][2] - BUDGET != 174008:
        raise Reject("adjacent excess")
    if EXPECTED[101155][-1] >= BUDGET:
        raise Reject("endpoint budget")
    return 41


def main() -> None:
    checks = validate(copy.deepcopy(EXPECTED))
    mutations = []
    for e, index, delta in ((98232, 0, 1), (101150, 6, -1),
                            (101155, 7, -1), (101156, 2, -1)):
        changed = copy.deepcopy(EXPECTED)
        row = list(changed[e])
        row[index] += delta
        changed[e] = tuple(row)
        try:
            validate(changed)
        except Reject:
            mutations.append(True)
        else:
            mutations.append(False)
    if not all(mutations):
        raise AssertionError("mutation controls")
    print(
        "MCA_FULL_LIFT_FIXED_CUTOFF_BOUNDARY_STACK_V1_PASS "
        f"checks={checks} mutations={sum(mutations)}/{len(mutations)}"
    )


if __name__ == "__main__":
    main()
